feat(networking): initial range based gets implementation

joshuef opened the pull request:
This pull request includes significant updates to the GitHub Actions workflows and several enhancements and fixes in the `sn_client` codebase. The changes improve the reliability and functionality of the workflows and add new methods and error handling to the client code.

### Workflow Improvements:

* [`.github/workflows/merge.yml`](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L768-R768): Updated the `log_file_prefix` and restructured the wallet creation and funding steps to include cleanup and verification processes. [[1]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L768-R768) [[2]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L894-R925)
* [`.github/workflows/merge.yml`](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1R1136-R1137): Added environment variables and a sleep step to ensure proper setup before running subsequent steps. [[1]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1R1136-R1137) [[2]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1R1214-R1227)
* [`.github/workflows/merge.yml`](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L1207-R1254): Improved directory checks and ensured no leftover files before proceeding with tests. [[1]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L1207-R1254) [[2]](diffhunk://#diff-bdb37b2ad65e049a0fc043e88813fdcf7ca0118abc11d99bbda676adfcbbb5a1L1227-R1267)

### Client Enhancements:

* [`sn_client/src/api.rs`](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L17-R25): Introduced a new method `get_range` and updated existing methods to include verification flags and enhanced error handling. [[1]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L17-R25) [[2]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97R309-R313) [[3]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L439-R497)
* [`sn_client/src/api.rs`](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L774-R821): Modified quorum settings and retry strategies for network operations to improve reliability. [[1]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L774-R821) [[2]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L808-R855) [[3]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97R895-R921)

### Error Handling Improvements:

* [`sn_client/src/error.rs`](diffhunk://#diff-a49580c567e54fb263fcfe3ab913094367d70f0a12b3681f9c65c197856e987dR16): Added new error types for faucet disbursement and no cash note found scenarios. [[1]](diffhunk://#diff-a49580c567e54fb263fcfe3ab913094367d70f0a12b3681f9c65c197856e987dR16) [[2]](diffhunk://#diff-a49580c567e54fb263fcfe3ab913094367d70f0a12b3681f9c65c197856e987dR27-R28) [[3]](diffhunk://#diff-a49580c567e54fb263fcfe3ab913094367d70f0a12b3681f9c65c197856e987dR51-R53)
* [`sn_client/src/faucet.rs`](diffhunk://#diff-e2aab3de22953c7d7077a1164e3957ef9b0c70d3631e8a44357f5c4e372609f9L41-R55): Implemented retry logic for loading cash notes from the faucet wallet.

### Register Handling:

* [`sn_client/src/register.rs`](diffhunk://#diff-1b86a48cb70322a3ac64029081787ac6c92b95fe1878ae05e65327b224bc787cL772-R774): Updated methods to use the new verification flag when retrieving signed registers from the network. [[1]](diffhunk://#diff-1b86a48cb70322a3ac64029081787ac6c92b95fe1878ae05e65327b224bc787cL772-R774) [[2]](diffhunk://#diff-1b86a48cb70322a3ac64029081787ac6c92b95fe1878ae05e65327b224bc787cL865-R869)

### Test Utilities:

* [`sn_client/src/test_utils.rs`](diffhunk://#diff-0e7d4cd94adcddff8cbd9e20764d29321aadb7695c473e24f560a8d67fbf8fd7L32-R32): Increased the number of retries for loading the faucet wallet in tests.

### Wallet Enhancements:

* [`sn_client/src/wallet.rs`](diffhunk://#diff-d655bcffabb264727633e45f79aff0b7990bb9045b3742352968837f26adca30L1012-R1012): Refined error handling for scenarios where all redeemed cash notes are already spent.

b-zee commented:
bors try

joshuef reviewed (approved):
bors r+

joshuef commented:
bors try-

joshuef committed:
feat(networking): initial range based gets implementation

joshuef committed:
chore(networking): rename sort_peers fns to clarify the accepted limit arg

joshuef committed:
chore(networking): remove REPLICATION_PEER_COUNT

joshuef committed:
fix(networking): reset raw spend get to be Majority

joshuef committed:
fix(networking): return true for exceeded quorum for non sensitive data

joshuef committed:
fix(networking): use cloest_node_buffer_zone when selecting nodes for range replication

joshuef committed:
fix(networking): use libp2p for register retrieval

merge the results after we see split record.
Use Quorum::Majority at least (instead of Quorum::One)

maqi commented on sn_client/src/api.rs:
```diff
@@ -436,18 +442,59 @@ impl Client {
+            PrettyPrintRecordKey::from(&record.key)
+        );
 
-        let maybe_records = self.network.get_register_record_from_network(key).await?;
```
I think we shall stay with the current `get_closest` first, then fetch from individual and merge.
swich back to use `kad::get_record` could be problematic, as explained in the slack discussion.

will continue the discussion within the slack discussion thread.

also, this seems not related to this change, and better to be in a separate PR ?

joshuef replied:
This is required for tests to pass, as the prior version does not work well with this. (We have no retry or majority of responses required... GetRange is naturally more probabilistic and has removed certain premature optimisation from the `get_record` calls which necessitate this I think.

`retry` there is inherently useful as we can hit more holders (now we have a potentially larger range).

So I think this approach is both simpler (`get_record` logic remains the same throughout calls to it), and healthier (going off the CI with/without)

maqi commented on sn_client/src/api.rs:
```diff
@@ -849,17 +896,25 @@ impl Client {
         // When there is retry on Put side, no need to have a retry on Get
         let verification_cfg = GetRecordCfg {
-            get_quorum: Quorum::Majority,
+            get_quorum: Quorum::All,
```
having an ALL with a bit shaky network could slow down or even block the uploader process.
as there will always be one node failes the get verification.
Any particular reason HAVE TO swtich back to ALL ?

same to couple of other places

joshuef replied:
> as there will always be one node failes the get verification

Why? 

Right now with `GetRange` all == `CloseGroupSize` as opposed to all returners... so it allows for stricter early exit.

Really though it's not so much about `All` so much as there being a difference in the checking between `verify` or not... So it could be `majority` in both cases and we just remove the `verify` flag on the fn perhaps? (Which I do note is as main... but it seemed handy to have this distinction to me...)

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -1050,23 +1048,40 @@ impl Client {
             let address = SpendAddress::from_unique_pubkey(spend.unique_pubkey());
-            debug!(
-                "Getting spend for pubkey {:?} from network at {address:?}",
+            warn!(
```
does it need to be as a `warn` ?
`info` shall be enough?

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -1050,23 +1048,40 @@ impl Client {
+                Ok(spend) => Ok(spend),
+                Err(error) => match error {
+                    Error::Network(sn_networking::NetworkError::DoubleSpendAttempt(spends)) => {
+                        warn!("DoubleSpentAttempt found with {spends:?}");
```
we shall change to use `BurntSpend` wherever we can ?

maqi commented on sn_networking/src/cmd.rs:
```diff
@@ -543,6 +556,31 @@ impl SwarmDriver {
     }
+
+    /// Return the RecordType
+    pub(crate) fn get_type_from_record(record: &Record) -> Result<RecordType> {
```
this seems shall be a helper function to be defined in the sn_transfer::Spend ?
or at least shall not be inside driver ?

joshuef replied:
It's unrelated to transfers. moved to protocol 👍

maqi commented on sn_networking/src/cmd.rs:
```diff
@@ -946,24 +965,72 @@ impl SwarmDriver {
+        target_address: &NetworkAddress,
+    ) -> Vec<PeerId> {
+        let filtered_peers = self.get_filtered_peers_exceeding_range(target_address);
+        let cloest_node_buffer_zone = CLOSE_GROUP_SIZE + close_group_majority();
```
cloest -> closest

maqi commented on sn_networking/src/driver.rs:
```diff
@@ -777,28 +789,25 @@ impl SwarmDriver {
                     // otherwise we're rewriting match statements etc around this anwyay
                     if let Err(err) = self.handle_swarm_events(swarm_event) {
-                        warn!("Error while handling swarm event: {err}");
+                        trace!("Issue while handling swarm event: {err}");
```
why lower the log level here?
if there is too frequent error message, it might mean there is mis-use ?
we shall mute those too frequent error by `match handle individually`, instead of lower log level for all ?

maqi commented on sn_networking/src/driver.rs:
```diff
@@ -777,28 +789,25 @@ impl SwarmDriver {
-                            self.replication_fetcher.set_replication_distance_range(distance);
-                        }
+                        let get_range = self.get_request_range();
+                        // set any new distance to farthest record in the store
```
the range is no longer the `farthest record` ?

maqi commented on sn_networking/src/driver.rs:
```diff
@@ -810,32 +819,94 @@ impl SwarmDriver {
+                }
+            }
+        }
+        all_peers.push(self.self_peer_id);
```
the function name says `get_all_local_peers_excluding_self` ?

maqi commented on sn_node/src/replication.rs:
```diff
@@ -76,12 +79,27 @@ impl Node {
+
+                                Record {
+                                    key,
+                                    value: bytes.to_vec(),
```
this means seralize two spends into one Record?
the `try_serialize_records`, and the `get_spend_from_record` functions are all shall get refactored to accomodate this.

Also, say one reocord is having content [1,2], and the other is with content [1], then a get_record will result with `SplitRecord` with both, then how these two got merged ?

joshuef replied:
> then how these two got merged ?

That's for node layer logic outwith of the get function now. (So the register apis eg, or the spend validation steps which already do this)

joshuef replied:
> this means seralize two spends into one Record?

This is already how this is stored, so there should be no change here. We're just ensuring that we do this on replication now where we did not previously (see `get_record_from_network` handling of double spend)

maqi commented on sn_networking/src/event/kad.rs:
```diff
@@ -369,92 +382,98 @@ impl SwarmDriver {
+    /// Checks passed peers from a request and checks they are sufficiently spaced to
+    /// ensure we have searched enough of the network range as determined by our `get_range`
+    ///
+    /// We expect any conflicting records to have been reported prior ti this check,
```
prior ti this -> prior this

maqi commented on sn_networking/src/event/kad.rs:
```diff
@@ -369,92 +382,98 @@ impl SwarmDriver {
+        data_key_address: &NetworkAddress,
+        quorum: &Quorum,
+    ) -> bool {
+        warn!("Assessing search: range: {:?}, address: {data_key_address:?}, quorum required: {quorum:?}, peers_returned_count: {:?}", expected_get_range.ilog2(), searched_peers_list.len());
```
info level ?

maqi commented on sn_networking/src/event/kad.rs:
```diff
@@ -369,92 +382,98 @@ impl SwarmDriver {
-                // Remove the query task and consume the variables.
-                let (_key, senders, result_map, _) = entry.remove();
+        // iterate over peers and see if the distance to the data is greater than the get_range
+        // Fathest peer from the data that has returned it
```
`Fathest peer from the data that has returned it`
this sentence can now be removed ?

joshuef committed:
chore(protocol): move get_type_from_record to protocol

joshuef committed:
fix(networking): remove self from get_peers_excluding_self

joshuef committed:
chore: small refactors and doc tweaks

joshuef committed:
fix(networking): remove custom REPLICATION_FACTOR

this indirectly affects how many close nodes returned and could stymie
the GetRange effectiveness

joshuef committed:
ci: normalize send amount for tests

joshuef committed:
ci: fixes

joshuef committed:
fix(client): retry faucet loading if genesis is spent

joshuef committed:
fix(faucet): loop if initial faucet balance 0 but genesis spent

joshuef committed:
fix(networking): ensure we update getrange when we can

ie, even if query times out if we have peers, we update

seems like we timeout more due to high K_VALUE

joshuef committed:
fix(networking): redial peers on intermittent errors

previously we just dropped any attempt at comms there

joshuef committed:
fix(networking): clarify faucet error and update test

joshuef committed:
ci: clarify duplicate genesis tests in merge.yml

joshuef committed:
fix(networking): local network dont force cleanup bootstrap

We were cleaning up bootstrap even if not full just because of being local

joshuef committed:
ci: split out cashnote check for clarity

joshuef committed:
test(networking): clarify that _at least_ all in range keys should be fetched

allow more to be fetched if needed due to distance variation

joshuef committed:
ci: assert faucet funded before upload

joshuef committed:
chore(net): remove clone post-rebase

joshuef committed:
ci: wait for faucet completion in dist test

joshuef committed:
chore(networking): small fixes comment updates

joshuef committed:
chore: post-rebase update get register call

joshuef committed:
fix: prevent recursive bootstrap quereis on peer addition

joshuef committed:
fix(nodeman): allow not found errors during cleanup
